### DISCUSSION TITLE
Material Properties through CSV

### DISCUSSION THREAD

**Discussion by manikumd22** [idaholab/moose#26355] [Q&A General]
I am new to MOOSE and C++, I am struggling with the problem like reading the CSV file and then using the values in the material. I am reading strain CSV file and I am using it to calculate _S[_qp] every gauss point. Is it correct way to write it. Please let me know
```
#### Reading the CSV file to get _S[_qp]

A = Strain.xlsx['Sheet1']
A1 = Strain.xlsx['Sheet2']


## Mathemtics to calculate _S[_qp]
B = A*_F[_qp]
_S[_qp] = A1*B

This is what I want to, and thus i define the the params

  params.addParam<MaterialPropertyName>("strain_excel","This strain will be used to calculate the stress tensor");

and then 
_strain_excel(getMaterialProperty<Real>(getParam<MaterialPropertyName>("strain_excel")))


But I don't know how to incorporate this in the calculation in the code. If someone please help me with this, it would be great


InputParameters
ComputeNeoHookeanStress::validParams()
{
  InputParameters params = ComputeLagrangianStressPK2::validParams();

  params.addParam<MaterialPropertyName>("lambda",
                                        "lambda",
                                        "Parameter conjugate to Lame parameter"
                                        " for small deformations");
  params.addParam<MaterialPropertyName>("mu",
                                        "mu",
                                        "Parameter conjugate to Lame parameter"
                                        " for small deformations");

  params.addParam<MaterialPropertyName>("strain_csv","This strain will be used to calculate the stress tensor");

  return params;
}

ComputeNeoHookeanStress::ComputeNeoHookeanStress(const InputParameters & parameters)
  : ComputeLagrangianStressPK2(parameters),
    _lambda(getMaterialProperty<Real>(getParam<MaterialPropertyName>("lambda"))),
    _mu(getMaterialProperty<Real>(getParam<MaterialPropertyName>("mu")))
    _strain_csv(getMaterialProperty<Real>(getParam<MaterialPropertyName>("strain_csv")))

{
}

void
ComputeNeoHookeanStress::computeQpPK2Stress()
{
  // Hyperelasticity is weird, we need to branch on the type of update if we
  // want a truly linear model
  //
  // This is because we need to drop quadratic terms for the linear update
  usingTensorIndices(i_, j_, k_, l_);

  // Large deformation = nonlinear strain
  if (_large_kinematics)
  {
    RankTwoTensor Cinv = (2 * _E[_qp] + RankTwoTensor::Identity()).inverse();
    _S[_qp] = (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv +
              _mu[_qp] * RankTwoTensor::Identity();
    _C[_qp] =
        -2 * (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv.times<i_, k_, l_, j_>(Cinv) +
        _lambda[_qp] * Cinv.times<i_, j_, k_, l_>(Cinv);
  }
  // Small deformations = linear strain
  else
  {
    const auto I = RankTwoTensor::Identity();
    RankTwoTensor strain = 0.5 * (_F[_qp] + _F[_qp].transpose()) - I;
    _C[_qp] = _lambda[_qp] * I.times<i_, j_, k_, l_>(I) +
              2.0 * _mu[_qp] * RankFourTensor(RankFourTensor::initIdentitySymmetricFour);
    
### Mathemtics defined above

    strain_PBC = _strain_csv
    _S[_qp] = strain_PBC;
  }
}

```

**Comment by GiudGiud** [comment]
Hello

If you just want to load quantities from CSV, you have to decide first what granularity you want for the data in the CSV. 

Per element? Per block? Per quadrature point?

Then you can either load the CSV with a function (PropertyFileReader then PiecewiseConstantFromCSV) or using our DelimitedFileReader utility directly in your material

Guillaume

**Comment by manikumd22** [comment]
Hi @GiudGiud 

At every gauss point, it will do some calculation but the data in Excel file is constant for the whole iteration, at every gauss point _F[_qp] is varying and based on that it will carry out calculation

A = Strain.xlsx['Sheet1']
A1 = Strain.xlsx['Sheet2']

Mathemtics to calculate _S[_qp]
B = A*_F[_qp]
_S[_qp] = A1*B

**Reply by GiudGiud** [reply]
ok that works. 
what is this
```
A = Strain.xlsx['Sheet1']
```
? Is this how you do it in Python?

Do you have the data tabulated by quadrature point right now?

**Reply by manikumd22** [reply]
I want to calculate_S[_qp] using the mathematics described below, instead of this _S[_qp] = _C[_qp] * strain

I have write this in python as I work on python,
```
import pandas as pd
import numpy as np

# Apply activation functions to each intermediate result
def relu_activation(x):
    return np.maximum(0, x)  # ReLU activation function



X_normalized = _F[_qp]


bias_weights_dict = pd.read_excel('biases_weights_data.xlsx', sheet_name=None)


weights_layer1 = bias_weights_dict['weight1'].values
weights_layer2 = bias_weights_dict['weight2'].values
weights_layer3 = bias_weights_dict['weight3'].values
weights_layer6 = bias_weights_dict['weight6'].values


# Access specific columns from the selected sheet
bias_fc1 = bias_weights_dict['bias1'].values
bias_fc2 = bias_weights_dict['bias2'].values
bias_fc3 = bias_weights_dict['bias3'].values
bias_fc6 = bias_weights_dict['bias6'].values



# Perform matrix multiplications

S = np.dot(weights_layer1, X_normalized.T) + bias_fc1.reshape(-1, 1)  # Add bias for fc1
S1 = np.dot(weights_layer2, S) + bias_fc2.reshape(-1, 1)  # Add bias for fc2
S2 = np.dot(weights_layer3, S1) + bias_fc3.reshape(-1, 1)  # Add bias for fc3



S = relu_activation(S)
S1 = relu_activation(S1)
S2 = relu_activation(S2)

_S[_qp] = np.dot(weights_layer6, S2) + bias_fc6.reshape(-1, 1)
```

This is what want to do?

How can I do this

 
Thanks

**Reply by GiudGiud** [reply]
you ll have to re-code this in C++ if you want to do this calculation dynamically in your moose simulation (and not just pre-compute a quantity in python, save it to csv then read it in C++)
I see relu in there, we also have an integration with libtorch

**Reply by manikumd22** [reply]
Hi @GiudGiud 

I started with modifying the ComputerNeoHookean.C by inserting the CSV file, but I am getting an error. Can you please help with this. I want that _lambda_csv will read value from CSV file although it is constant for whole mesh but it will read through CSV file
Error:
```
*** ERROR ***

[31m
*** ERROR ***
basic_string::substr: __pos (which is 18446744073709551615) > this->size() (which is 4)[39m

```


I am attaching the CSV file, input file, .C file and .h file

E_values.csv:

E
1.10E+06


.i file:

```
[Functions]
  [element]
    type = PiecewiseConstantFromCSV
    read_prop_user_object = 'reader_element'
    read_type = 'element'  # Adjust based on your data setup
    column_number = '1'    # Assuming column 1 contains the lambda values
  []
[]

[UserObjects]
  [reader_element]
    type = PropertyReadFile
    prop_file_name = 'E_values.csv'  # Adjust this to your CSV file path
    read_type = 'element'  # Adjust based on your data setup
    nprop = 1  # Update this according to the number of columns in your CSV
  []
[]

  [Materials]
    [compute_stress]
      type = ComputeNeoHookeanStress
      _lambda_csv = element

      mu = 8300.33333888888926
    []
    [compute_strain]
      type = ComputeLagrangianStrain
    []
  []


  

  .C file:

#include "ComputeNeoHookeanStress.h"

registerMooseObject("TensorMechanicsApp", ComputeNeoHookeanStress);

InputParameters
ComputeNeoHookeanStress::validParams()
{
  InputParameters params = ComputeLagrangianStressPK2::validParams();

  params.addParam<MaterialPropertyName>("_lambda_csv",
                                        "_lambda_csv",
                                        "Parameter conjugate to Lame parameter"
                                        " for small deformations");
  params.addParam<MaterialPropertyName>("mu",
                                        "mu",
                                        "Parameter conjugate to Lame parameter"
                                        " for small deformations");
  return params;
}

ComputeNeoHookeanStress::ComputeNeoHookeanStress(const InputParameters & parameters)
  : ComputeLagrangianStressPK2(parameters),
    _lambda_csv(getMaterialProperty<Real>(getParam<MaterialPropertyName>("_lambda_csv"))),
    _mu(getMaterialProperty<Real>(getParam<MaterialPropertyName>("mu")))

{
}

void
ComputeNeoHookeanStress::computeQpPK2Stress()
{
  // Hyperelasticity is weird, we need to branch on the type of update if we
  // want a truly linear model
  //
  // This is because we need to drop quadratic terms for the linear update
  usingTensorIndices(i_, j_, k_, l_);

  // Large deformation = nonlinear strain
  if (_large_kinematics)
  {
    RankTwoTensor Cinv = (2 * _E[_qp] + RankTwoTensor::Identity()).inverse();
    _S[_qp] = 2.0*(_lambda_csv[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv +
              _mu[_qp] * RankTwoTensor::Identity();
    _C[_qp] =
        2.0 * (_lambda_csv[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv.times<i_, k_, l_, j_>(Cinv) +
        _lambda_csv[_qp] * Cinv.times<i_, j_, k_, l_>(Cinv);
  }
  // Small deformations = linear strain
  else
  {
    const auto I = RankTwoTensor::Identity();
    RankTwoTensor strain = 0.5 * (_F[_qp] + _F[_qp].transpose()) - I;
    _C[_qp] = 2.0*_lambda_csv[_qp] * I.times<i_, j_, k_, l_>(I) +
              2.0 * _mu[_qp] * RankFourTensor(RankFourTensor::initIdentitySymmetricFour);
    _S[_qp] = _C[_qp] * strain;
  }
}
  




.h file:

#pragma once

#include "ComputeLagrangianStressPK2.h"

/// Compressible Neo-Hookean hyperelasticity
///
///  Model follows from W = lambda / 2 * (ln J)^2 - mu * ln J + 1/2 * mu *
///  (tr(C)- I)
///
///  with C = 1/2*(F.T*F-I) and J = det(F)
///
class ComputeNeoHookeanStress : public ComputeLagrangianStressPK2
{
public:
  static InputParameters validParams();
  ComputeNeoHookeanStress(const InputParameters & parameters);

protected:
  /// Actual stress/Jacobian update
  virtual void computeQpPK2Stress();

protected:
  const MaterialProperty<Real> & _lambda_csv;
  const MaterialProperty<Real> & _mu;
};

```

**Reply by GiudGiud** [reply]
> basic_string::substr: __pos (which is 18446744073709551615) > this->size() (which is 4)�[39m

this is an out of bounds access. If you compile in debug mode (METHOD=debug make -j6) you will get a line for this problem. Then you can fix that line

**Reply by manikumd22** [reply]
Hi @GiudGiud 
I compiled METHOD=debug make -j6 and then I tried to run my code. But I'm getting same error


I think I am facing an error because of this


[Functions]
  [element]
    type = PiecewiseConstantFromCSV
    read_prop_user_object = 'reader_element'
    read_type = 'element'  # Adjust based on your data setup
    column_number = '1'    # Assuming column 1 contains the lambda values
  []
[]

[UserObjects]
  [reader_element]
    type = PropertyReadFile
    prop_file_name = 'E_values.csv'  # Adjust this to your CSV file path
    read_type = 'element'  # Adjust based on your data setup
    nprop = 1  # Update this according to the number of columns in your CSV
  []
[]



Is it the correct way to import CSV file and in my CSV file there is just one column that corresponds to the E value with just 1 value.

E
1.10E+06

**Reply by GiudGiud** [reply]
No your data is not tabulated by element so the PropertyFileReader is not well set up. 

The data_type is probably block/subdomain not element

**Reply by manikumd22** [reply]
Hi @GiudGiud 

Can you please tell me how to do it?

I tried with block but still same error.


Can you tell me what should be the format of csv file?

```
 [Functions]
    [block]
      type = PiecewiseConstantFromCSV
      read_prop_user_object = 'reader_block'
      read_type = 'block'
      column_number = '1'
    []
  []

  [UserObjects]
    [reader_block]
      type = PropertyReadFile
      prop_file_name = 'E_values.csv'
      read_type = 'block'
      nprop = 1
      nblock = 1
    []
  []

  [Materials]
    [compute_stress]
      type = ComputeNeoHookeanStress
      _lambda_csv = 'block'
      mu = 8300.33333888888926
    []
    [compute_strain]
      type = ComputeLagrangianStrain
    []
  []
```

**Reply by GiudGiud** [reply]
How many blocks do you have in the simulation?

How many entries do you have in your file?

**Reply by manikumd22** [reply]
Hi @GiudGiud 

1. I am not sure what does block means, if it means domain then for my whole domain, there is one block
2. Right now there would be only one column (E values) to start with, but as I progresses I will modify

**Reply by GiudGiud** [reply]
A block is the same a subdomain. In paraivew you can see blocks in the left panel, they are selectable just like nodesets

One column (number of properties to read from file) but how many rows (subdomains for which you have data for)

**Reply by manikumd22** [reply]
Hi @GiudGiud 

I checked

id Blocks
5 Surface

Rows:
1. Right now, I have just one row for the whole domain, but later I want to include many rows (say 64) for whole domain

**Reply by GiudGiud** [reply]
from the documentation
```
!alert warning title=Data ordering
For the `element` (`block` and `node` respectively) modes, both the elements
(blocks and nodes respectively) in the mesh and the data in the `CSV` file must be ordered consecutively.
The element (block and node respectively) IDs must be contiguous and usually start from 1.
```

so you should make 5 rows of data, 4 dummy, one real, if you want your first block to have an id of 5

**Reply by manikumd22** [reply]
Hi @GiudGiud 

I tried with this CSV file and getting an error
1, 0.0    # Dummy data for blocks 1-4
2, 0.0
3, 0.0
4, 0.0
5, 10.0   # Actual data for block 5

*** ERROR ***
Failed to convert a delimited data into double when reading line 1 in file /storage/home/mvk6158/work/moose_directories/mooseApp/E_values.csv.
  LINE 1: , 0.0    # Dummy data for blocks 1-4[39m

**Reply by GiudGiud** [reply]
try this, with no comments
```
1, 0.0
2, 0.0
3, 0.0
4, 0.0
5, 10.0
```

**Reply by manikumd22** [reply]
Hi @GiudGiud 

I tried and getting error

*** ERROR ***
*** ERROR ***
The following error occurred in the object "reader_block", of type "PropertyReadFile".

Row 0 in /storage/home/mvk6158/work/moose_directories/mooseApp/E_values.csv has number of data less than 2�[39m

application called MPI_Abort(MPI_COMM_WORLD, 1) - process 1

I am assigning in this way, Is may way is correct?

[Functions]
[./block]
type = PiecewiseConstantFromCSV
read_prop_user_object = 'reader_block'
read_type = 'block'
column_number = '2'. # as I have 2nd columns for numeric value
[../]
[]

[UserObjects]
[reader_block]
type = PropertyReadFile
prop_file_name = 'E_values.csv'
read_type = 'block'
nprop = 2 # as I have 2 columns
nblock = 1 # as I have 1 block
[]
[]

[Materials]
[compute_stress]
type = ComputeNeoHookeanStress
_lambda_csv = 'block'
mu = 8300.33333888888926
[]
[compute_strain]
type = ComputeLagrangianStrain
[]
[]

**Reply by GiudGiud** [reply]
```
nprop = 2 # as I have 2 columns
```
nprop = 1 I think, one of the column is just an index.

**Reply by manikumd22** [reply]
Hi @GiudGiud 

I tried and getting this error

*** ERROR ***
The following error occurred in the object "MOOSE Problem", of type "FEProblem".

Material property 'block', requested by 'compute_stress' is not defined on block surface
Material property 'block', requested by 'compute_stress_face' is not defined on block surface
Material property 'block', requested by 'compute_stress_neighbor' is not defined on block surface

**Reply by GiudGiud** [reply]
yup you have now moved on to the next error. The CSV file is being read properly.

You need to look at which mateiral propery you called 'block' and forgot to declare as 'block'

**Reply by manikumd22** [reply]
Hi @GiudGiud 
Oh god, thanks!

I think, I am declaring the block and then assigning  _lambda_csv  = 'block'? If not, can you please tell me how to modify this

but I am  
```
[Functions]
    [block]
      type = PiecewiseConstantFromCSV
      read_prop_user_object = 'reader_block'
      read_type = 'block'
      column_number = '2'
      property = _lambda_csv
    []
  []

  [UserObjects]
    [reader_block]
      type = PropertyReadFile
      prop_file_name = 'E_values.csv'
      read_type = 'block'
      nprop = 1
      nblock = 1
    []
  []

  [Materials]
    [compute_stress]
      type = ComputeNeoHookeanStress
      _lambda_csv = 'block'
      mu = 8300.33333888888926
    []
    [compute_strain]
      type = ComputeLagrangianStrain
    []
  []
```

**Reply by GiudGiud** [reply]
`block` is a function here (which btw, maybe rename it! It can get confusing to use that name)
```
      _lambda_csv = 'block'
```
expects a material property 
(unless you have modified from 
```
_lambda(getMaterialProperty(getParam("lambda"))),
```

so you can use the function directly here.
Instead, use this:
https://mooseframework.inl.gov/source/materials/GenericFunctionMaterial.html
to go from a function to a material property

**Reply by manikumd22** [reply]
Hi @GiudGiud 

I changed, but still getting an error
The following error occurred in the object "MOOSE Problem", of type "FEProblem".
```
Material property 'lambda', requested by 'compute_stress' is not defined on block surface
Material property 'lambda', requested by 'compute_stress_face' is not defined on block surface
Material property 'lambda', requested by 'compute_stress_neighbor' is not defined on block surface





 [Functions]
    [./_lambda_csv]
      type = PiecewiseConstantFromCSV
      read_prop_user_object = 'reader_block'
      read_type = 'block'
      column_number = '2'
    [../]
  []

  [Materials]
    [compute_stress]
      type = ComputeNeoHookeanStress
      prop_names = lambda
      prop_values = _lambda_csv
      mu = 8300.33333888888926
    []
    [compute_strain]
      type = ComputeLagrangianStrain
    []
  []
```

.C 
```
  params.addParam<MaterialPropertyName>("lambda",
                                        "lambda",
                                        "Parameter conjugate to Lame parameter"
                                        " for small deformations");

**Reply by GiudGiud** [reply]
```
    [compute_stress]
      type = ComputeNeoHookeanStress
      prop_names = lambda
      prop_values = _lambda_csv
      mu = 8300.33333888888926
    []
```
this looks wrong. Do you mean to use a GenericFunctionMaterial for the type?

**Reply by manikumd22** [reply]
Hi @GiudGiud 

Because I want to lambda to take values from CSV file rather than defining a numeric value

If I changed it to  GenericFunctionMaterial then how will I do this calculation, because I want to ComputeNeoHookeanStress

ComputeNeoHookeanStress:


ComputeNeoHookeanStress::computeQpPK2Stress()
{
  // Hyperelasticity is weird, we need to branch on the type of update if we
  // want a truly linear model
  //
  // This is because we need to drop quadratic terms for the linear update
  usingTensorIndices(i_, j_, k_, l_);

  // Large deformation = nonlinear strain
  if (_large_kinematics)
  {
    RankTwoTensor Cinv = (2 * _E[_qp] + RankTwoTensor::Identity()).inverse();
    _S[_qp] = 2.0*(_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv +
              _mu[_qp] * RankTwoTensor::Identity();
    _C[_qp] =
        2.0 * (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv.times<i_, k_, l_, j_>(Cinv) +
        _lambda[_qp] * Cinv.times<i_, j_, k_, l_>(Cinv);
  }
  // Small deformations = linear strain
  else
  {
    const auto I = RankTwoTensor::Identity();
    RankTwoTensor strain = 0.5 * (_F[_qp] + _F[_qp].transpose()) - I;
    _C[_qp] = 2.0*_lambda[_qp] * I.times<i_, j_, k_, l_>(I) +
              2.0 * _mu[_qp] * RankFourTensor(RankFourTensor::initIdentitySymmetricFour);
    _S[_qp] = _C[_qp] * strain;
  }
}





For example, in one of the given examples, they are varying youngs_modulus for type PiecewiseLinearInterpolationMaterial

[Materials]
  [./youngs_modulus]
    type = PiecewiseLinearInterpolationMaterial
    x = '100 500'
    y = '1e6 6e5'
    property = youngs_modulus
    variable = temp
  [../]
  [./poissons_ratio]
    type = PiecewiseLinearInterpolationMaterial
    x = '100 500'
    y = '0   0.25'
    property = poissons_ratio
    variable = temp
  [../]

  [./elasticity_tensor]
    type = ComputeVariableIsotropicElasticityTensor
    args = temp
    youngs_modulus = youngs_modulus
    poissons_ratio = poissons_ratio
  [../]

**Reply by GiudGiud** [reply]
You can define two materials. One for the stress, one for forwarding the function to a material property which you can then use in NeoHookean stress

**Reply by manikumd22** [reply]
Hi @GiudGiud 

I define the Generic Function Material, is it right way to do define it and how to call this function in ComputeNeoHookeanStress (lambda_csv is the martial property name)
```
[Materials]
    [./a]
      type = GenericFunctionMaterial
      block = 5
      prop_names = lambda_csv
      prop_values = lambda1
    [../]
    [compute_stress]
      type = ComputeNeoHookeanStress
      lambda_csv = ????
      mu = 8300.33333888888926
    []


[Functions]
    [./lambda1]
      type = PiecewiseConstantFromCSV
      read_prop_user_object = 'reader_block'
      read_type = 'block'
      column_number = '1'
    [../]
  []

  [UserObjects]
    [reader_block]
      type = PropertyReadFile
      prop_file_name = 'E_values.csv'
      read_type = 'block'
     nprop = 1
      nblock = 5
    []
  []
```

**Reply by GiudGiud** [reply]
try this
```
    [compute_stress]
      type = ComputeNeoHookeanStress
      lambda_csv = lambda_csv
      mu = 8300.33333888888926
    []
```

since lambda_csv is the name that you are using here for the material property:
```
    [./a]
      type = GenericFunctionMaterial
      block = 5
      prop_names = lambda_csv #<- that one
      prop_values = lambda1
    [../]
```

**Reply by manikumd22** [reply]
Hi @GiudGiud 

I tried, but I got an error

Error:

*** ERROR ***
/storage/home/mvk6158/work/moose_directories/mooseApp/manik_neohookean_small.i:222: (UserObjects/reader_block/nblock):
    Element block id 5 greater than than total number of blocks in mesh: 5. Blocks should be numbered consecutively.[39m

application called MPI_Abort(MPI_COMM_WORLD, 1) - process 2




 [Functions]
    [./lambda1]
      type = PiecewiseConstantFromCSV
      read_prop_user_object = 'reader_block'
      read_type = 'block'
      column_number = '1'
    [../]
  []

  [UserObjects]
    [reader_block]
      type = PropertyReadFile
      prop_file_name = 'E_values.csv'
      read_type = 'block'
      nprop = 1
      nblock = 5
    []
  []

  [Materials]
    [compute_stress]
      type = ComputeNeoHookeanStress
      lambda_csv = lambda_csv
      mu = 8300.33333888888926
    []
    [./a]
      type = GenericFunctionMaterial
      block = 5
      prop_names = lambda_csv
      prop_values = lambda1
    [../]
    [compute_strain]
      type = ComputeLagrangianStrain
    []
  []

**Reply by GiudGiud** [reply]
That s good you moved on to the next error. 

You need to recognize when you are making progress otherwise you will end up getting stuck breaking things you have already gotten right

For this next one, you should comply with what the error says. 

Use a renameSubdomaiGenerator in the [Mesh] block to rename you block from 5 to 0

**Reply by manikumd22** [reply]
Hi @GiudGiud 

It is working now, thank you so much for your time and effort. 

Previously, I was generating mesh with Gmesh, I have changed as per one of the 2D mesh example given in MOOSE in which there is only one block whose id starts from 0.


Although I have solved this issue, when I change the values I get same result, it's not changing at all
```
One value of lambda_csv = 20000000000000000.0
0, 20000000000000000.0

Other value of lambda_csv = 0.0
0, 0.0
```

But when I enter the numeric value, the results are changing

It means it's not reading the correct value
```
 [Functions]
    [./lambda1]
      type = PiecewiseConstantFromCSV
      read_prop_user_object = 'reader_block'
      read_type = 'block'
      column_number = '1'
    [../]
  []
  [UserObjects]
    [reader_block]
      type = PropertyReadFile
      prop_file_name = 'E_values.csv'
      read_type = 'block'
      nprop = 1
      nblock = 1
    []
  []

  [Materials]
    [compute_stress]
      type = ComputeNeoHookeanStress
      lambda_csv = lambda_csv
      mu = 6700.0
    []
    [./a]
    type = GenericFunctionMaterial
    block = 0
    prop_names = lambda_csv
    prop_values = lambda1
    [../]
    [compute_strain]
      type = ComputeLagrangianStrain
    []
  []




// Large deformation = nonlinear strain
if (_large_kinematics)
{
RankTwoTensor Cinv = (2 * _E[_qp] + RankTwoTensor::Identity()).inverse();
const Real lambda = _lambda_csv[_qp];
_S[_qp] = lambda*(log(_F[_qp].det()) - _mu[_qp]) * Cinv +
lambda*mu[qp] * RankTwoTensor::Identity();
C[qp] =
-2 * lambda*(log(F[qp].det()) - mu[qp]) * Cinv.times<i, k, l, j>(Cinv) +
lambda * Cinv.times<i, j, k, l>(Cinv);

}
// Small deformations = linear strain
else
{
const auto I = RankTwoTensor::Identity();
RankTwoTensor strain = 0.5 * (_F[_qp] + _F[_qp].transpose()) - I;
const Real lambda = _lambda_csv[_qp];

C[qp] = 2.0*lambda * I.times<i, j, k, l>(I) +
lambda * 2.0 * _mu[_qp] * RankFourTensor(RankFourTensor::initIdentitySymmetricFour);
_S[_qp] = _C[_qp] * strain;
}
}
```

**Reply by GiudGiud** [reply]
Well that's odd. I dont know what could be happening so we'd likely have to debug it to find out why it s not updating

**Reply by mvkpenn22** [reply]
Hi @GiudGiud 

I tried with this modification, and it worked. I set column_number = '0' instead of 1. I don't know the reason, how it worked. Can you tell me what this column_number represents?

    [./lambda1]
      type = PiecewiseConstantFromCSV
      read_prop_user_object = 'reader_block'
      read_type = 'block'
      column_number = '0'
    [../]

**Reply by GiudGiud** [reply]
it s the column index into the CSV file. I m guessing there is an offset then. We ll document it
